Sprint2

j-hellenberg opened the pull request.

Margarete01 committed:
add DictionarySegment files, start implementing compress_chunk function

Commit:
changes to compress method

dey4ss committed:
Sprint 2 (#3)

j-hellenberg committed:
Move draft of compression implementation to DictionarySegment

j-hellenberg committed:
First draft for constructing a dictonary segment from a value segment

Margarete01 committed:
implement dictionary segment methods

Margarete01 committed:
refactor dictionary segment methods

j-hellenberg committed:
Clean up dictionary segment construction

j-hellenberg committed:
Introduce nullValueId for dictionary segment and use uint_*s for the abstract attribute vector

Margarete01 committed:
add Dictionary Test WrongTypeCompression

Margarete01 committed:
address concurrent access and multithreading encoding

j-hellenberg committed:
Improve multithreaded compression and ensure the chunk access is locked properly during compression

j-hellenberg committed:
Correct memory use estimation for dictionary segment

j-hellenberg committed:
Use recursive_mutex instead of mutex to allow multiple access from the same thread

j-hellenberg committed:
Add some tests for the dictionary segment and the table

j-hellenberg committed:
Convert some DebugAsserts to Asserts because we also need them in the release build

phkeese commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -0,0 +1,153 @@
+
+template <typename T>
+DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(abstract_segment);
+  Assert(value_segment, "Can only construct a DictionarySegment from a value segment of matching type.");
```
Why are you enforcing a specific type of AbstractSegment here and not instead type-casting the individual values? This way, only `ValueSegment`s can be compressed.

j-hellenberg replied:
We decided to do it that way because the task hints that only value segments can be compressed (right now). This allows us to access the values vector of the segment, which we would not be able to otherwise (we also did not want to use the `operator[]` on the segment because it returns AllTypeVariants, which is inefficient to work with

phkeese commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -0,0 +1,153 @@
+  // We need to make sure to only put values in our dictionary that don't correspond to NULL_VALUE.
+  // However, we cannot just remove the default value for T (for example, "" for std::string) from our final dictionary,
+  // because somebody might have actually inserted this value without meaning the NULL_VALUE.
+  for (auto offset = ChunkOffset{0}; offset < original_values.size(); ++offset) {
```
Consider using another variable for `original_values.size()`

phkeese commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -0,0 +1,153 @@
+  if (_dictionary.size() > std::numeric_limits<uint16_t>::max()) {
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint32_t>>();
+    _null_value_id = INVALID_VALUE_ID;
+  } else if (_dictionary.size() > std::numeric_limits<uint8_t>::max()) {
```
Consider using another variable for `_dictionary.size()`

j-hellenberg replied:
I think it should be fine here because we are not calling it inside a loop.

phkeese commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -0,0 +1,153 @@
+    _null_value_id = static_cast<uint16_t>(INVALID_VALUE_ID);
+  } else {
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint8_t>>();
+    _null_value_id = static_cast<uint8_t>(INVALID_VALUE_ID);
```
How exactly do you differentiate between a NULL-value and an invalid value (e.g. upper bound)?

j-hellenberg replied:
Our reasoning was that they can be easily distinguished by context. When you do `lower_bound` you know that you can receive a `INVALID_VALUE_ID` (but not a `NULL_VALUE_ID`). You should not use that `valueID` to do `value_of_value_id`, as that ID will be invalid.

On the other hand, when calling, e.g., `get_typed_value`, you know that your attribute vector might contain a `NULL_VALUE_ID`, which you handle as being a `NULL_VALUE`

phkeese commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  const auto original_values = value_segment->values();
+  for (auto offset = ChunkOffset{0}; offset < original_values.size(); ++offset) {
```
Consider using another variable for `original_values.size()`

phkeese commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -0,0 +1,153 @@
+
+template <typename T>
+AllTypeVariant DictionarySegment<T>::operator[](const ChunkOffset chunk_offset) const {
+  return _dictionary.at(_attribute_vector->get(chunk_offset));
```
Consider using `Assert` instead of `at()` to make throwing checks explicit.
Also check other usages of `at()`.

phkeese commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -0,0 +1,153 @@
+ValueID DictionarySegment<T>::value_id_for_value(const T value) const {
+  const auto index = find(_dictionary.begin(), _dictionary.end(), value);
+  Assert(index != _dictionary.end(), "Value is not contained in dictionary.");
+  return static_cast<ValueID>(std::distance(_dictionary.begin(), index));
```
Consider returning `INVALID_VALUE_ID` in case of a missing value instead of throwing. This would allow users of this method to use it for checking wether a value is contained within the dictionary.

j-hellenberg replied:
Nice point. Considered it, but came to the conclusion that, as this method is only used internally during segment construction, this is not really needed and would even complicate things because we would then be forced to check that the return value of this method is not `INVALID_VALUE_ID`. I added a comment on this instead :)

phkeese commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -0,0 +1,153 @@
+
+template <typename T>
+size_t DictionarySegment<T>::estimate_memory_usage() const {
+  return _attribute_vector->width() * size() + sizeof(T) * _dictionary.size();
```
As mentioned in class, please consider using `capacity()` instead of `size()` to get an accurate memory consumption:

```suggestion
  return _attribute_vector->width() * size() + sizeof(T) * _dictionary.capacity();
```

phkeese commented on src/lib/storage/dictionary_segment.hpp:
```diff
@@ -0,0 +1,76 @@
+  // Returns the ValueID representing a certain value.
+  ValueID value_id_for_value(const T value) const;
+  void construct_dictionary(const std::shared_ptr<ValueSegment<T>>& value_segment);
+  void construct_attribute_vector(const std::shared_ptr<ValueSegment<T>>& value_segment);
```
Please review the style guide and mark protected and private members with an underscore:

```suggestion
  ValueID _value_id_for_value(const T value) const;
  void _construct_dictionary(const std::shared_ptr<ValueSegment<T>>& value_segment);
  void _construct_attribute_vector(const std::shared_ptr<ValueSegment<T>>& value_segment);
```

phkeese commented on src/lib/storage/dictionary_segment.hpp:
```diff
@@ -0,0 +1,76 @@
+
+class AbstractAttributeVector;
+
+// Dictionary is a specific segment type that stores all its values in a vector
```
Not your fault, but sentences should end in full stop:
```suggestion
// Dictionary is a specific segment type that stores all its values in a vector.
```

phkeese commented on src/lib/storage/fixed_width_integer_vector.cpp:
```diff
@@ -0,0 +1,31 @@
+namespace opossum {
+template <typename T>
+ValueID FixedWidthIntegerVector<T>::get(const size_t index) const {
+  return static_cast<ValueID>(_values.at(index));
```
Consider using `Assert` for explicit bounds checks.

phkeese commented on src/lib/storage/fixed_width_integer_vector.cpp:
```diff
@@ -0,0 +1,31 @@
+  if (_values.size() <= index) {
+    _values.resize(index + 1);
+  }
+  _values[index] = value_id;
```
Consider using `Assert` to test wether `value_id` can be represented using `T` to defend against too large indices and unexpected behavior.
Also think about what happens when you insert a large index and have to resize and fill with indices.

j-hellenberg replied:
We decided to follow your example and construct our `fixed_with_integer_vector` directly from a vector instead of using `set` which requires us to resize the vector for every value we enter

phkeese commented on src/lib/storage/table.cpp:
```diff
@@ -100,17 +102,54 @@ bool Table::column_nullable(const ColumnID column_id) const {
+  const auto segment = chunk_to_be_compressed->get_segment(index);
+  resolve_data_type(column_type(index), [&new_chunk, &segment](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
+    new_chunk->add_segment(std::make_shared<DictionarySegment<ColumnDataType>>(segment));
```
Possible race condition between segment ordering: 
Using `new_chunk->add_segment()` blindly may insert a segment out-of-order if its thread finished faster than another thread. Consider passing a segment index to each thread and writing into a vector and then finally merging that vector into a new chunk.
Also: Your implementation of `Chunk::add_segment()` uses `vector::emplace_back` which is [not thread-safe](https://en.cppreference.com/w/cpp/container#Thread_safety).

phkeese commented on src/lib/storage/table.cpp:
```diff
@@ -100,17 +102,54 @@ bool Table::column_nullable(const ColumnID column_id) const {
+  // We cannot hand out this chunk if compression is currently in progress, as somebody might modify it,
+  // causing us to lose this inserted data once the compression (which was based on the old chunk state) has finished.
+  // If the caller can ensure he won't modify the chunk, he can use the const overload instead.
+  const std::lock_guard<std::recursive_mutex> lock(_chunk_access_lock);
   return _chunks[chunk_id];
```
This blocks access to all chunks, not just currently compressing chunks.
Also, this prevents compressing multiple chunks at the same time.

Not your fault, but:
Users of this function may choose to retain a copy of this pointer and later access it while it is being compressed.
This could lead to some segments containing rows, other segments do not. Again, not your fault but worth mentioning.

j-hellenberg replied:
Yeah, so apparently, this was a futile attempt to address the requirement of the task document.
Decided to remove this lock altogether and instead add a comment on why replacing the chunks should work fine without data races even without the lock.

phkeese commented on src/lib/storage/table.cpp:
```diff
@@ -100,17 +102,54 @@ bool Table::column_nullable(const ColumnID column_id) const {
+  const auto chunk_to_be_compressed = get_chunk(chunk_id);
+
+  std::vector<std::thread> compression_threads;
+  for (auto index = ColumnID{0}; index < column_count(); index++) {
```
Consider using another variable for `column_count()`.

phkeese commented on src/lib/storage/table.cpp:
```diff
@@ -100,17 +102,54 @@ bool Table::column_nullable(const ColumnID column_id) const {
+    thread.join();
+  }
+
+  _chunks.at(chunk_id) = new_chunk;
```
Consider using `operator[]` here, since you already do bounds checking earlier in this method.

phkeese commented on src/lib/storage/table.hpp:
```diff
@@ -68,10 +69,13 @@ class Table : private Noncopyable {
   std::vector<std::string> _column_names;
   std::vector<std::string> _column_types;
   std::vector<bool> _column_nullables;
+  void compress_segment_and_add_to_chunk(ColumnID index, const std::shared_ptr<Chunk>& new_chunk,
```
Please prepend protected members with `_`:
```suggestion
  void _compress_segment_and_add_to_chunk(ColumnID index, const std::shared_ptr<Chunk>& new_chunk,
```

phkeese commented on src/test/storage/dictionary_segment_test.cpp:
```diff
@@ -0,0 +1,152 @@
+  EXPECT_EQ(dict_segment->estimate_memory_usage(), size_t{135});
+}
+
+// TODO(student): You should add some more tests here (full coverage would be appreciated) and possibly in other files.
```
```suggestion

```

phkeese commented on src/test/storage/dictionary_segment_test.cpp:
```diff
@@ -0,0 +1,152 @@
+  value_segment_str->append(2);
+  auto dict_segment = std::make_shared<DictionarySegment<std::string>>(value_segment_str);
+
+  EXPECT_EQ(dict_segment->estimate_memory_usage(), size_t{99});
```
Consider writing a comment for how this magic number comes to be.
Two more cases below.

phkeese reviewed (changes requested):
Nicely done, though you should take another look at your locking mechanism and particularly segment ordering during compression.

-- Team 2 @phkeese @ClFeSc @23mafi

j-hellenberg committed:
Address review comments

j-hellenberg committed:
Remove locking mechanism for preventing chunk access during compression.
This is because we cannot fully prevent modifications to the chunk using that method anyway,
but the edge case in which our mechanism would have been required violated the inteded usage pattern anyway.

j-hellenberg committed:
In the memory usage tests for the dictionary segment, replace the magic values with something more meaningful
